core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -271,56 +247,82 @@ private boolean unsubscribe(String queryName,
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(sqm -> filter.test((SubscriptionQueryMessage<?, ?, U>) sqm))
+                              .map(m -> (SubscriptionQueryMessage<?, ?, U>) m)
+                              .forEach(query -> {
+                                  UpdateHandler<?, U> updateHandler = (UpdateHandler<?, U>) updateHandlers.get(query);
+                                  try {
+                                      U calculatedUpdate = update.apply(query);
+                                      updateHandler.onUpdate(calculatedUpdate);
+                                  } catch (Exception e) {
+                                      updateHandlers.remove(query);
+                                      updateHandler.onCompletedExceptionally(e);
+                                  }
+                              }));
+    }
+
+    @Override
+    public void complete(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(filter)
+                              .forEach(query -> {
+                                  UpdateHandler<?, ?> updateHandler = updateHandlers.get(query);
+                                  updateHandlers.remove(query);
+                                  updateHandler.onCompleted();
+                              }));
+    }
+
+    @Override
+    public void completeExceptionally(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter, Throwable cause) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(filter)
+                              .forEach(query -> {
+                                  UpdateHandler<?, ?> updateHandler = updateHandlers.get(query);
+                                  updateHandlers.remove(query);
+                                  updateHandler.onCompletedExceptionally(cause);
+                              }));
+    }
+
+    @SuppressWarnings("unchecked")
+    private void registeringSubscriptionQueryReadSafe(Predicate<?> filter, Runnable r) {
+        List<ReentrantReadWriteLock> locks = subscriptionQueriesRegistering.keySet()
+                                                                           .stream()
+                                                                           .filter((Predicate<? super SubscriptionQueryMessage<?, ?, ?>>) filter)
+                                                                           .map(subscriptionQueriesRegistering::get)
+                                                                           .collect(Collectors.toList());
+
+        locks.forEach(lock -> lock.readLock().lock());
```

Comment:
This will wait until all subscription queries matching given filter are registered, and after that will execute the given emit/complete/completeExceptionally call. It is crucial that *all* queries are registered before we execute given operation (we don't want to have a query which misses the operation).